_posts/2019-09-30-introducing-arrow-flight.md

```diff
@@ -0,0 +1,278 @@
+clients and servers to send data and metadata to each other simultaneously
+while requests are being served.
+
+A simple Flight setup might consist of a single server to which clients connect
+and make `DoGet` requests.
+
+<div align="center">
+<img src="{{ site.baseurl }}/img/20191014_flight_simple.png"
+     alt="Flight Simple Architecture"
+     width="50%" class="img-responsive">
+</div>
+
+## Optimizing Data Throughput over gRPC
+
+While using a general-purpose messaging library like gRPC has numerous specific
+benefits beyond the obvious ones (taking advantage of all the engineering that
+Google has done on the problem), some work was needed to improve the
+performance of transporting large datasets. Many kinds of gRPC users only deal
+with relatively small messages, for example.
+
+The best-supported way to use gRPC is to define services in a [Protocol
+Buffers][3] (aka "Protobuf") `.proto` file. A Protobuf plugin for gRPC
+generates gRPC service stubs that you can use to implement your
+applications. RPC commands and data messages are serialized using the [Protobuf
+wire format][4]. Because we use "vanilla gRPC and Protocol Buffers", gRPC
+clients that are ignorant of the Arrow columnar format can still interact with
+Flight services and handle the Arrow data opaquely.
+
+The main data-related Protobuf type in Flight is called `FlightData`. Reading
+and writing Protobuf messages in general is not free, so we implemented some
+low-level optimizations in gRPC in both C++ and Java to do the following:
+
+* Generate the Protobuf wire format for `FlightData` including the Arrow record
+  batch being sent without going through any intermediate memory copying or
+  serialization steps.
+* Reconstruct a Arrow record batch from the Protobuf representation of
+  `FlightData` without any memory copying or deserialization. In fact, we
+  intercept the encoded data payloads without allowing the Protocol Buffers
+  library to touch them.
+
+In a sense we are "having our cake and eating it, too". Flight implementations
+having these optimizations will have better performance, while naive gRPC
+clients can still talk to the Flight service and use a Protobuf library to
+deserialize `FlightData` (albeit with some performance penalty).
+
+As far as absolute speed, in our C++ data throughput benchmarks, we are seeing
+end-to-end TCP throughput in excess of 2-3GB/s on localhost without TLS
+enabled. This benchmark shows a transfer of ~12 gigabytes of data in about 4
+seconds:
+
+
+```shell
+$ ./arrow-flight-benchmark --records_per_stream 100000000
+Bytes read: 12800000000
+Nanos: 3900466413
+Speed: 3129.63 MB/s
+```
+
+From this we can conclude that the machinery of Flight and gRPC adds relatively
+little overhead, and it suggests that many real-world applications of Flight
+will be bottlenecked on network bandwidth.
+
+## Horizontal Scalability: Parallel and Partitioned Data Access
```

Comment:
I think we should put at least one sentence about parallel access in the top of the blog entry. It is one of the key differentiators of the flight approach versus several traditional approaches and is a bit buried at the moment.

Reply:
done